### PR TITLE
i#2941 recreate_app_state spill/restore: test restore of non-public DR slot in mangling

### PR DESCRIPTION
**Pull request by hgreving2304**
Adds test to check whether drreg ignores restoring non-public slot of DR used for mangling. DR secretly violates transparency in extreme cases if scratch reg is optimized into using app's dead reg of rip-rel instruction. This patch adds an option to DR used by this test to avoid this.

Issue: #2941, #2933

### CODE REVIEW

**Comment by hgreving2304** [comment]
FYI even w/ PTRSZ, the MSVS assembler is going to have a build error once the test will get re-enabled after i#3312. Technically, there is a second Win issue that I can't reproduce locally, only AppVeyor.

File: suite/tests/client-interface/drreg-test.c [hunk 1]
```diff
@@ -179,6 +205,21 @@ main(int argc, const char *argv[])
         test_asm_faultD();
     }
 
+#    if defined(UNIX)
+    intercept_signal(SIGSEGV, (handler_3_t)&handle_signal4, false);
+#    elif defined(WINDOWS)
+    SetUnhandledExceptionFilter(&handle_exception4);
+#    endif
+
+    /* Test 10: fault test restore of non-public DR slot used by mangling,
```
Comment (derekbruening) [inline suite/tests/client-interface/drreg-test.c]: "test fault restore", right?

File: suite/tests/client-interface/drreg-test.c [hunk 2]
```diff
@@ -494,6 +535,56 @@ GLOBAL_LABEL(FUNCNAME:)
...
+#ifdef X86
+#ifdef X64
+        PUSH_CALLEE_SAVED_REGS()
+        sub      REG_XSP, FRAME_PADDING /* align */
+        END_PROLOG
+
+          /* XXX i#3312: Temporarily disable test until bug has been fixed. */
+#if 0
+        jmp      test10
+        /* Test 10: fault test restore of non-public DR slot used by mangling,
```
Comment (derekbruening) [inline suite/tests/client-interface/drreg-test.c]: test fault

File: suite/tests/client-interface/drreg-test.c [hunk 3]
```diff
@@ -494,6 +535,56 @@ GLOBAL_LABEL(FUNCNAME:)
...
+        sub      REG_XSP, 8
+        mov      TEST_REG_ASM, DRREG_TEST_10_ASM
+        mov      TEST_REG_ASM, DRREG_TEST_10_ASM
+        nop
+        mov      REG_XAX, DRREG_TEST_11_ASM
+        /* There's a good chance that this will be too far away for a
+         * signed 32-bit rip-rel offset, which is what we want, in order
+         * to get the address mangled into register REG_XAX.
+         */
+        mov      REG_XAX, PTRSZ SYMREF(-0x7fffffff) /* crash */
```
Comment (derekbruening) [inline suite/tests/client-interface/drreg-test.c]: How about just making this an add that reads xax too and then there is no need for a special option?

File: core/arch/x86/mangle.c [hunk 4]
```diff
@@ -2885,7 +2885,8 @@ mangle_rel_addr(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
                 relop = instr_get_src(instr, si);
                 ASSERT(di < 0 || opnd_same(relop, instr_get_dst(instr, di)));
                 /* If it's a load (OP_mov_ld, or OP_movzx, etc.), use dead reg */
-                if (instr_num_srcs(instr) == 1 && /* src is the rip-rel opnd */
+                if (DYNAMO_OPTION(mangle_rel_addr_scratch_reg_opt) &&
```
[the remark this answers is not in the document]
Comment (hgreving2304) [reply core/arch/x86/mangle.c]: See below

File: suite/tests/client-interface/drreg-test.c (hunk 2 above)
Comment (hgreving2304) [reply suite/tests/client-interface/drreg-test.c]: Yes

File: suite/tests/client-interface/drreg-test.c (hunk 1 above)
Comment (hgreving2304) [reply suite/tests/client-interface/drreg-test.c]: Ack

File: suite/tests/client-interface/drreg-test.c (hunk 3 above)
Comment (hgreving2304) [reply suite/tests/client-interface/drreg-test.c]: You're right. Done.

**Pull request merged**